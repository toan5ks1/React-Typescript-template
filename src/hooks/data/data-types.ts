import {
    InfiniteQueryObserverResult,
    QueryKey,
    UseQueryOptions as rqUseQueryOptions,
    UseQueryResult,
} from "react-query";
import { ProviderTypes } from "src/common/constants/enum";
import { TypeEntity } from "src/typings/react-admin";

export enum UseQueryOptions {
    RETRY_TIMES = 0,
    CACHE_TIME = 0,
    TIME_UNTIL_AUTO_REFETCH = 5 * 60 * 1000, // don't auto reload, auto fetch for 5 min after the query executed
}

export type ActionTypeForQuery =
    | ProviderTypes.ONE
    | ProviderTypes.LIST
    | ProviderTypes.MANY
    | ProviderTypes.TITLE
    | ProviderTypes.GET_TITLES
    | ProviderTypes.COUNT_STATUS
    | ProviderTypes.DISPLAY_ORDER
    | ProviderTypes.GET_USERNAMES
    | ProviderTypes.DOWNLOAD_BOOK
    | ProviderTypes.MANY_REFERENCE
    | ProviderTypes.LIST_WITH_FILTER
    | ProviderTypes.RETRIEVE_GRADE_MAP
    | ProviderTypes.DOWNLOAD_STUDY_PLAN
    | ProviderTypes.VALIDATE_EXTERNAL_ID
    | ProviderTypes.GET_STATUS_NOTIFICATION
    | ProviderTypes.COUNT_READ_OF_NOTIFICATIONS
    | ProviderTypes.GET_STUDENTS_WITH_SCHOOL
    | ProviderTypes.GET_GRADES_OF_STUDENTS
    | ProviderTypes.GET_BRIGHTCOVE_VIDEO_INFO
    | ProviderTypes.COUNT_SCHOOL_ADMINS
    | ProviderTypes.GET_PREVIOUS_LESSON_REPORT
    | ProviderTypes.RETRIEVE_PARTNER_DOMAIN
    | ProviderTypes.GET_ATTENDANCE_STATUS_AND_REMARK
    | ProviderTypes.PARTNER_DYNAMIC_FORM_FIELD_VALUES_WITH_LESSON_REPORT_DETAILS
    | ProviderTypes.ONE_LATEST
    | ProviderTypes.MANY_REFERENCE_WITHOUT_CALC
    | ProviderTypes.MANY_REFERENCE_WITH_LOCATION_WITHOUT_CALC
    | ProviderTypes.MANY_WITH_LOCATION;

export interface GraphqlPagination {
    limit?: number;
    offset?: number;
}

export interface LegacyGraphqlPagination {
    page?: number;
    perPage?: number;
}

export interface PayloadForQuery<Sort extends Record<any, any> = any> {
    filter?: Record<string, any>;
    sort?: Sort;
    pagination?: GraphqlPagination | LegacyGraphqlPagination;
}

export interface UseQueryBaseRequest<Sort extends Record<any, any> = any> {
    entity: TypeEntity;
    action: ActionTypeForQuery;
    params?: PayloadForQuery<Sort>;
}

export type ProviderResponse<T> = {
    data: T;
    total?: number;
};

export interface UseQueryBaseOptions<T = any, MappedType = T, IError = Error> {
    retry?: number;
    enabled?: boolean;
    cacheFor?: number; // ms to cache the result
    markDataAsStaleAfter?: number;
    rerenderOnChangeProps?: Array<keyof InfiniteQueryObserverResult>; // will only rerender when those props change, use with care

    selector?: rqUseQueryOptions<T, IError, MappedType>["select"]; // select/filter/transform to get data you want based on returned result
    onError?: rqUseQueryOptions<T, IError, MappedType>["onError"];
    onSuccess?: rqUseQueryOptions<T, IError, MappedType>["onSuccess"];
}

export type UseQueryBaseV2Return<T> = UseQueryResult<T, Error> & {
    queryKey: QueryKey | undefined;
};

export function getQueryKey<Order>(params: UseQueryBaseRequest<Order>) {
    return `${params.entity}_${params.action}`;
}
