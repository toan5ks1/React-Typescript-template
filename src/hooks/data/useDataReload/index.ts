export { default } from "./useDataReload";
export * from "./useDataReload";
