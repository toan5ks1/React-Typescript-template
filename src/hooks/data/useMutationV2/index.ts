export { default } from "./useMutationV2";
export * from "./useMutationV2";
