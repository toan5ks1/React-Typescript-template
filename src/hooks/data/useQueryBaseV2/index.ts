export { default } from "./useQueryBaseV2";
export * from "./useQueryBaseV2";
