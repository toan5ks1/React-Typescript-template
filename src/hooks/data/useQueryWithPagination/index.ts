export { default } from "./useQueryWithPagination";
export * from "./useQueryWithPagination";
