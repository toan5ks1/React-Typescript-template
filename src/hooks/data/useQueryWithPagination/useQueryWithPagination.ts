import { useEffect, useMemo, useRef } from "react";

import { DataProviderResult, UseQueryValue, useSafeSetState, useVersion } from "react-admin";
import { usePrevious } from "react-use";
import { NotifyTypes } from "src/common/constants/enum";
import { handleUnknownError } from "src/common/utils/error";
import { RaQuery } from "src/typings/react-admin";

import useDataProvider from "../../useDataProvider";
import useShowSnackbar from "../../useShowSnackbar";
import usePagination, { UsePaginationOptions, UsePaginationReturn } from "../usePagination";

import isEqual from "lodash/isEqual";

export interface PaginationWithTotal extends Omit<UsePaginationReturn, "setLimit" | "setOffset"> {
    count: number;
}

export interface PaginationReturn {
    result: UseQueryValue;
    pagination: PaginationWithTotal;
}

function mergePagination(
    obj: RaQuery,
    pagination: Omit<UsePaginationReturn, "setOffset" | "setLimit">
) {
    return {
        ...obj,
        payload: {
            ...obj.payload,
            pagination: {
                page: pagination.page + 1, // material use zero-based, but hasura is 1-based
                perPage: pagination?.limit || 0,
            },
        },
    };
}

function useQueryWithPagination(
    query: RaQuery,
    paginationOptions?: UsePaginationOptions
): PaginationReturn {
    const { setOffset, setLimit, ...pagination } = usePagination(paginationOptions);
    const provider = useDataProvider();
    const { resource, payload, type } = mergePagination(query, pagination);
    const prevPayload = usePrevious(payload);
    const [result, safeSetResult] = useSafeSetState<UseQueryValue>({
        data: [],
        error: null,
        total: undefined,
        loading: true,
        loaded: false,
        refetch: () => {},
    });
    const version = useVersion();
    const prevVersion = usePrevious(version);
    const requestCount = useRef(0);
    const showSnackbar = useShowSnackbar();

    useEffect(() => {
        if (isEqual(prevPayload, payload) && prevVersion === version) {
            return;
        }

        const { filter, pagination, sort } = payload;

        // Prevent two simultaneous requests from happening
        // when querying on a page other than 1
        if (!isEqual(prevPayload?.filter, filter)) {
            pagination.page = 1;
            setOffset(0);
        }

        safeSetResult((result) => ({ ...result, loading: true, loaded: false }));

        const currentCount = ++requestCount.current;

        (async () => {
            try {
                const result: DataProviderResult = await provider(type, resource, {
                    filter,
                    pagination,
                    sort,
                });

                // Prevent race condition
                if (currentCount < requestCount.current) return;

                if (!Array.isArray(result.data)) result.data = [];

                safeSetResult({ ...result, loading: false, loaded: true, refetch: () => {} });
            } catch (err) {
                const error = handleUnknownError(err);
                safeSetResult({ data: [], error, loading: false, loaded: true, refetch: () => {} });
                showSnackbar(error.message, NotifyTypes.ERROR);
            }
        })();
    }, [
        prevPayload,
        query,
        provider,
        resource,
        payload,
        type,
        version,
        prevVersion,
        safeSetResult,
        setOffset,
        showSnackbar,
    ]);

    const paginationWithTotal = useMemo(() => {
        return {
            ...pagination,
            count: typeof result.total === "number" ? result.total : 0,
        };
    }, [result, pagination]);

    return {
        result,
        pagination: paginationWithTotal,
    } as const;
}

export default useQueryWithPagination;
