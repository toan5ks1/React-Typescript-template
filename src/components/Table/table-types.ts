import { ComponentProps, CSSProperties, ElementType, HTMLAttributes, Key, ReactNode } from "react";

import { StandardProps } from "src/typings/react-component";

import { Table, TableCellProps } from "@mui/material";

import { PaginationWithTotal } from "src/hooks/data/useQueryWithPagination";

export interface CellProps extends TableCellProps {
    "data-testid"?: string;
    style?: CSSProperties;
    headerStyle?: CSSProperties;
}

export interface TableColumn<T extends object = any> {
    key: string;
    title?: ReactNode;
    emptyValue?: string;
    testId?: string;
    cellProps?: CellProps;
    dataIndex?: string;
    render?: (record: T, dataIndex?: number) => ReactNode;
}

export type TableIndex = boolean | Pick<CSSProperties, "width">;

export enum TableBorderStyle {
    all = "all",
    horizontal = "horizontal",
    none = "none",
}

export interface RowKeyFunc {
    (record: object): string;
}

export type StyleKeys = {
    root: CSSProperties;
    container: CSSProperties;
    table: CSSProperties;
    row: CSSProperties;
    header: CSSProperties;
    footer: CSSProperties;
};

export interface TableCheckboxSelectType {
    onSelectAllClick: (isSelectAll: boolean) => void;
    hasData: boolean;
    hasChecked: boolean;
    isCheckedAll: boolean;
}

export interface TableBaseHeaderProps extends StandardProps {
    visible: boolean;
    border: CSSProperties;
    columns: TableColumn[];
    withIndex?: TableIndex;
    render?: (columns: TableColumn[]) => ReactNode;
}

export interface TableBaseBodyProps<TData>
    extends Pick<
        TableBaseRowProps<TData>,
        "border" | "rowKey" | "columns" | "hover" | "withIndex"
    > {
    data: TData[];
    loading: boolean;
    noDataMessage?: string | ReactNode;
    isSelected?: Function;
    customBody?: ReactNode;
    skeCount?: number;
    skeHeight?: number;
    pagination?: PaginationWithTotal;
    styleIndexCol?: TableBaseRowProps<TData>["styleIndexCol"];
}

export interface TableBaseRowProps<TData>
    extends Omit<HTMLAttributes<HTMLTableRowElement>, "onSelect"> {
    dataIndex: number;
    record: TData;
    columns: TableColumn[];
    selected?: boolean;
    rowKey: string | RowKeyFunc;
    hover?: boolean;
    border: CSSProperties;
    withIndex?: TableIndex;
    pagination?: PaginationWithTotal;
    render?: (columns: TableColumn[]) => ReactNode;
    styleIndexCol?: CSSProperties;
}

export interface TableBaseCellProps extends CellProps {
    border?: Pick<
        CSSProperties,
        "border" | "borderLeft" | "borderTop" | "borderRight" | "borderBottom"
    >;
    maxColumnIndex?: number;
    columnIndex?: number;
}

export interface TableBaseFooterProps {
    pagination: PaginationWithTotal;
    labelRowsPerPage?: string;
    rowsPerPageOptions?: number[];
    style?: CSSProperties;
}

export interface TableBaseProps<TData> extends StandardProps {
    columns: TableColumn[];
    data: TData[];
    showHeader: boolean;
    border: keyof typeof TableBorderStyle; // Support 3 border style
    body: Omit<TableBaseBodyProps<TData>, "data" | "columns" | "border">;
    footer?: TableBaseFooterProps;
    styles?: Partial<StyleKeys>; // Support custom style by using style attribute
    tableProps?: object | ComponentProps<typeof Table>; // Sticky header, put data-id...
    component: ElementType;
    isSelected?: (key: Key) => boolean;
    errorMessage?: string | ReactNode;
    withIndex?: TableIndex;
    styleIndexCol?: TableBaseRowProps<TData>["styleIndexCol"];
}

export interface TableHeaderWithCheckboxProps extends Omit<TableBaseHeaderProps, "render"> {
    select?: TableCheckboxSelectType;
}

export interface TableBodyWithCheckboxProps<TData>
    extends TableBaseBodyProps<TData>,
        Pick<TableRowWithCheckboxProps<TData>, "onSelect" | "handleClickCheckbox">,
        Pick<TableWithCheckboxProps<TData>, "hasDraftProperty"> {}

export interface TableRowWithCheckboxProps<TData>
    extends Omit<TableBaseRowProps<TData>, "render">,
        Pick<TableWithCheckboxProps<TData>, "hasDraftProperty"> {
    handleClickCheckbox?: (record: TData) => void;
    onSelect?: (record: TData[]) => void;
}

export interface TableWithCheckboxProps<TData> extends Omit<TableBaseProps<TData>, "isSelected"> {
    listSelectedItems?: TData[]; // Allow pass a list with selected item from outside
    onSelect?: (record: TData[]) => void; // Support checkbox each row
    hasDraftProperty?: boolean;
}

export interface TableBodyWithRowSpanProps<TData> extends Omit<TableBaseBodyProps<TData>, "data"> {
    data: { [group: string]: TData[] };
}

export interface TableWithRowSpanProps<TData>
    extends Omit<TableBaseProps<TData>, "columns" | "data"> {
    headerColumns: TableColumn[];
    dataColumns: TableColumn[];
    data: { [group: string]: TData[] };
}
