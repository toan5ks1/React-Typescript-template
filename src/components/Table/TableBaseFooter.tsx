import { TablePagination, LabelDisplayedRowsArgs } from "@mui/material";
import { inputBaseClasses } from "@mui/material/InputBase";
import { tablePaginationClasses } from "@mui/material/TablePagination";
import TypographyBase from "src/components/Typographys/TypographyBase";

import { TableBaseFooterProps } from "./table-types";

import useTranslate from "src/hooks/useTranslate";

const sx = {
    [`& .${inputBaseClasses.input}`]: {
        padding: "9px 35px 7px 8px",
        fontSize: "12px",
    },
    [`& .${tablePaginationClasses.selectIcon}`]: {
        top: "calc(50% - 13px)",
        right: "7px",
    },
    [`& .${tablePaginationClasses.select}`]: {
        paddingRight: "35px !important",
    },
};

const TableBaseFooter = (props: TableBaseFooterProps) => {
    const t = useTranslate();
    const {
        pagination,
        labelRowsPerPage = t("ra.navigation.page_rows_per_page"),
        rowsPerPageOptions = [5, 10, 25, 50, 100],
        style,
    } = props;

    const paginationValue: LabelDisplayedRowsArgs = {
        count: pagination.count,
        from: pagination.page * pagination.rowsPerPage + 1,
        page: pagination.page,
        to: pagination.page * pagination.rowsPerPage + pagination.rowsPerPage,
    };

    const labelDisplayedRows =
        paginationValue.count > -1
            ? t("ra.navigation.page_range_info", {
                  offsetBegin: paginationValue.from,
                  offsetEnd: Math.min(paginationValue.to, paginationValue.count),
                  total: paginationValue.count,
              })
            : `${paginationValue.from}-${paginationValue.to}`;

    return (
        <TablePagination
            SelectProps={{
                id: "TableBaseFooter__select",
            }}
            backIconButtonProps={{
                size: "large",
            }}
            nextIconButtonProps={{
                size: "large",
            }}
            labelDisplayedRows={() => (
                <TypographyBase data-testid="TableBaseFooter__totalRows" variant="caption">
                    {labelDisplayedRows}
                </TypographyBase>
            )}
            data-testid="TableBaseFooter"
            {...pagination}
            labelRowsPerPage={<TypographyBase variant="caption">{labelRowsPerPage}</TypographyBase>}
            rowsPerPageOptions={rowsPerPageOptions}
            count={pagination?.count}
            sx={sx}
            component="div"
            style={style}
        />
    );
};

export default TableBaseFooter;
