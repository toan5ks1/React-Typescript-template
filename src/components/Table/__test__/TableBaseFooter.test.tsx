import { shallow } from "enzyme";

import TableBaseFooter from "../TableBaseFooter";
import { TableBaseFooterProps } from "../table-types";

import { render } from "@testing-library/react";
import { PaginationWithTotal } from "src/hooks/data/useQueryWithPagination";

describe("<TableBasePagination />", () => {
    const pagination: PaginationWithTotal = {
        count: 10,
        offset: 0,
        page: 0,
        rowsPerPage: 10,
        limit: 10,
        onPageChange: jest.fn(),
        onRowsPerPageChange: jest.fn(),
    };
    let props: TableBaseFooterProps = {
        pagination: pagination,
        labelRowsPerPage: "label",
        rowsPerPageOptions: [5, 10],
    };

    it("should match snapshot", () => {
        const wrapper = shallow(<TableBaseFooter {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render without count property", () => {
        const wrapper = render(<TableBaseFooter {...props} />);
        expect(wrapper.container.querySelector("input")?.value).toEqual(
            `${props.pagination.count}`
        );
    });
});
