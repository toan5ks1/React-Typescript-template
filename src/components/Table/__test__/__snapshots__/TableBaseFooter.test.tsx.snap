// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableBasePagination /> should match snapshot 1`] = `ShallowWrapper {}`;
